### PR TITLE
[GOBBLIN-182] Add lineage event for query based source

### PR DESCRIPTION
**Pull request by yukuai518**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-182

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):
    The initial version of lineage events for query based source.

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+ *  LINEAGE_PREFIX  TASK_PREFIX   K.2 ---> V.2
+ *  LINEAGE_PREFIX  TASK_PREFIX   K.3 ---> V.3
+ */
+@Slf4j
+public class LineageInfo {
+  public static final String PUBLISH_OUTPUT = "publish.output";
+  public static final String WRITER_OUTPUT = "writer.output";
+  public enum Level {
+    JOB_LEVEL, // only load job level (dataset level) lineage info
+    TASK_LEVEL, // load task level lineage info
```
Comment (ibuenros) [inline gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java]: Should we also have a `DATASET_LEVEL`?

File: gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+ *
+ * When serialize to {@link State} or {@link WorkUnit}, a LINEAGE_PREFIX prefix is added to help deserialization.
+ *
+ *  LINEAGE_PREFIX                K.1 ---> V.1
+ *  LINEAGE_PREFIX  TASK_PREFIX   K.2 ---> V.2
+ *  LINEAGE_PREFIX  TASK_PREFIX   K.3 ---> V.3
+ */
+@Slf4j
+public class LineageInfo {
+  public static final String PUBLISH_OUTPUT = "publish.output";
```
Comment (ibuenros) [inline gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java]: What are these used for?

File: gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+  @Getter
+  private String jobId;
+
+  private Map<String, String> lineageMetaData;
+
+  private LineageInfo() {
+  }
+
+  private LineageInfo(String datasetUrn, String jobId) {
+    Preconditions.checkState(datasetUrn != null);
```
Comment (ibuenros) [inline gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java]: `datasetUrn` is not a state, it is an argument. Same in next line.

File: gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+        if (level == Level.TASK_LEVEL) {
+          descriptor.lineageMetaData.put(strippedKey, value);
+        } else if (level == Level.JOB_LEVEL) {
+          if (!strippedKey.startsWith(TASK_PREFIX)) {
+            descriptor.lineageMetaData.put(strippedKey, value);
+          }
+        } else if (level == Level.TASK_LEVEL_ONLY) {
+          if (strippedKey.startsWith(TASK_PREFIX)) {
+            descriptor.lineageMetaData.put(strippedKey, value);
```
Comment (ibuenros) [inline gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java]: Shouldn't you also strip `TASK_PREFIX`?

File: gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+              jobLevel.get(strippedKey).add(value);
+            }
+          } else {
+            throw new UnsupportedOperationException("only job level aggregation is supported, will consider task level support int he future.");
+          }
+        }
+      }
+    }
+
+    for (Map.Entry<String, Set<String>> entry: jobLevel.entrySet()) {
```
Comment (ibuenros) [inline gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java]: If you are going to use only one, there is no point using a set above. You can just compare against the current value, and if it is different, throw an exception early at that point.

File: gobblin-core/src/main/java/org/apache/gobblin/publisher/BaseDataPublisher.java [hunk 6]
```diff
@@ -342,9 +346,16 @@ protected void publishData(WorkUnitState state, int branchId, boolean publishSin
 
       movePath(parallelRunner, state, writerOutputDir, publisherOutputDir, branchId);
       writerOutputPathsMoved.add(writerOutputDir);
+      addPublisherLineageInfo(state, branchId, publisherOutputDir.toString());
     }
   }
 
+  protected void addPublisherLineageInfo(WorkUnitState state, int branchId, String output) {
+    LineageInfo info = LineageInfo.load(state, LineageInfo.Level.TASK_LEVEL);
```
Comment (ibuenros) [inline gobblin-core/src/main/java/org/apache/gobblin/publisher/BaseDataPublisher.java]: Is there a need to deserialize the existing `LineageInfo`? Can't you just add the new stuff without deserializing the old stuff?

File: gobblin-core/src/main/java/org/apache/gobblin/writer/FsDataWriter.java [hunk 7]
```diff
@@ -247,6 +256,18 @@ public void commit()
...
+ }
+
+ protected void addWriteOutputLineageInfo (State state, int branchId, String output) {
+   LineageInfo info = LineageInfo.load(state, LineageInfo.Level.TASK_LEVEL);
+   info.setProp(Joiner.on(".").join(LineageInfo.WRITER_OUTPUT, branchId), output, LineageInfo.Level.TASK_LEVEL);
+   info.save(state);
+
+   if (state.contains(ConfigurationKeys.JOB_DATA_PUBLISHER_TYPE) && state.getProp(ConfigurationKeys.JOB_DATA_PUBLISHER_TYPE).equals(
+       NoopPublisher.class.getName())) {
+     submitter.submit(info.getUUID(), info.getLineageMetaData());
```
Comment (ibuenros) [inline gobblin-core/src/main/java/org/apache/gobblin/writer/FsDataWriter.java]: This seems a bit awkward here. You could refactor this into a separate class, or just specify that Noop publisher will not send lineage events.

File: gobblin-modules/gobblin-sql/src/main/java/org/apache/gobblin/source/extractor/extract/jdbc/MysqlSource.java [hunk 8]
```diff
@@ -50,4 +53,15 @@
...
     return extractor;
   }
+
+  public LineageInfo createLineageInfo (SourceEntity entity, SourceState state){
+    LineageInfo descriptor = super.createLineageInfo(entity, state);
+    String host = state.getProp(ConfigurationKeys.SOURCE_CONN_HOST_NAME);
+    String port = state.getProp(ConfigurationKeys.SOURCE_CONN_PORT);
+    String database = state.getProp(ConfigurationKeys.SOURCE_QUERYBASED_SCHEMA);
+    String connectionUrl = "jdbc:mysql://" + host.trim() + ":" + port + "/" + database.trim();
+    descriptor.setProp("connectionUrl", connectionUrl, LineageInfo.Level.JOB_LEVEL);
```
Comment (ibuenros) [inline gobblin-modules/gobblin-sql/src/main/java/org/apache/gobblin/source/extractor/extract/jdbc/MysqlSource.java]: Please make `connectionUrl` a constant.

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java [hunk 9]
```diff
@@ -169,6 +177,26 @@ public Void call()
...
+    Collection<TaskState> states = datasetState.getTaskStates();
+    if (states.size() == 0) {
+      return;
+    }
+
+    TaskState oneWorkUnitState = states.iterator().next();
+    if (oneWorkUnitState.contains(ConfigurationKeys.JOB_DATA_PUBLISHER_TYPE) && oneWorkUnitState.getProp(ConfigurationKeys.JOB_DATA_PUBLISHER_TYPE).equals(
+        NoopPublisher.class.getName())) {
+      // if no publisher configured, each task should be responsible for sending lineage event.
+      return;
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java]: Isn't it easier to do it here either way? What's the point of making the task do the work?

File: gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageException.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package org.apache.gobblin.lineage;
+
+public class LineageException extends Exception {
```
Comment (ibuenros) [inline gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageException.java]: Please add javadoc.

File: gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+
+  /**
+   * Get all lineage meta data.
+   */
+  public ImmutableMap<String, String> getLineageMetaData() {
+    return ImmutableMap.copyOf(lineageMetaData);
+  }
+
+
+  public static LineageInfo load (Collection<? extends State> states, Level level) throws LineageException {
```
Comment (ibuenros) [inline gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java]: Please add javadoc and specify that all states need to be part of the same branch / dataset depending on the level.

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java [hunk 12]
```diff
@@ -169,6 +177,29 @@ public Void call()
...
+    }
+
+    try {
+      LineageInfo info = LineageInfo.load(datasetState.getTaskStates(), LineageInfo.Level.All);
+      EventSubmitter submitter = new EventSubmitter.Builder(Instrumented.getMetricContext(datasetState, SafeDatasetCommit.class),
+          LineageInfo.LINEAGE_NAME_SPACE).build();
+      submitter.submit(info.getUUID(), info.getLineageMetaData());
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java]: Why is the name `info.getUUID`?

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java [hunk 13]
```diff
@@ -159,6 +164,9 @@ public Void call()
         } else if (canPersistStates) {
           persistDatasetState(datasetUrn, datasetState);
         }
+
+        submitLineageEvent(datasetState);
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java]: Is the event submitted even if the publishing failed?

File: gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+  }
+
+  /**
+   * Retrieve lineage information from a {@link State} by {@link Level}
+   * @param level {@link Level#DATASET}  only load dataset level lineage attributes
+   *              {@link Level#BRANCH}   only load branch level lineage attributes
+   *              {@link Level#All}      load all lineage attributes
+   * @return A {@link LineageInfo} object containing all lineage attributes
+   */
+  public static LineageInfo load (State state, Level level) throws LineageException {
```
Comment (ibuenros) [inline gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java]: Isn't this the same as `load(Collection, Level)` when the collection contains a single state?

File: gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+            metaData.put(lineageKey.substring(DATASET_PREFIX.length()), lineageValue);
+          }
+        }
+      }
+    }
+
+    String jobId = metaData.containsKey(ConfigurationKeys.JOB_ID_KEY)? metaData.get(ConfigurationKeys.JOB_ID_KEY):"";
+    String urn = metaData.containsKey(ConfigurationKeys.DATASET_URN_KEY)? metaData.get(ConfigurationKeys.DATASET_URN_KEY):ConfigurationKeys.DEFAULT_DATASET_URN;
+
+    metaData.remove(ConfigurationKeys.JOB_ID_KEY);
```
Comment (ibuenros) [inline gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java]: This requires the job id and dataset urn to exist twice in the state, once under `job.id` and `dataset.urn`, and again with the lineage prefix. Why not read it from the original key-value pairs.

File: gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java (hunk 14 above)
Comment (yukuai518) [inline gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java]: reuse collection version

File: gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java (hunk 15 above)
Comment (yukuai518) [inline gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java]: use original one

File: gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java (hunk 11 above)
Comment (yukuai518) [inline gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java]: done

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java (hunk 13 above)
Comment (yukuai518) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java]: done

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java (hunk 12 above)
Comment (yukuai518) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java]: Change to getId()

**Comment by ibuenros** [comment]
+1

**Pull request closed without merging**